### PR TITLE
1.17 implementation

### PR DESCRIPTION
**Pull request by dfsek**
# Pull Request

## Brief description.

Switches to Java 16 and switches the Fabric implementation to 1.17.

Temporarily disables the `:platforms:forge` subproject, since ForgeGradle is not compatible with Gradle 7.

## Licensing

- [X] I am the original author of this code, and I am willing to release it under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html).
- [ ] I am not the original author of this code, but it is in public domain or released
  under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html) or a compatible license.

## Goal of the PR

 - [x] Remove the few remaining assumptions of world height from `common`
 - [x] Add `material-overrides` key to ore configs, to make configuring ores such as deepslate ores easier.
 - [x] Update to 1.17

#### Types of changes

- [ ] Bug Fix 
- [ ] Build system 
- [ ] Documentation 
- [x] New Feature 
- [ ] Performance 
- [ ] Refactoring 
- [ ] Repository 
- [ ] Revert 
- [ ] Style 
- [ ] Tests 
- [ ] Translation 

#### Compatiblity

- [ ] Breaking change 
- [x] Non-Breaking change.

This does not break *any* config features. There is 1 optional config addition. The Fabric implementation will, however, no longer run on 1.16.x.

#### Contribution Guidelines.

- [x] I have read the [`CONTRIBUTING.md`](https://github.com/PolyhedralDev/Terra/blob/master/CONTRIBUTING.md) document in the root of the
  git repository.
- [x] My code follows the code style for this
  project.
#### Documentation

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

#### Testing

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: common/src/main/java/com/dfsek/terra/api/structures/script/builders/ZeroArgFunctionBuilder.java [hunk 1]
```diff
@@ -42,7 +42,7 @@ public Position getPosition() {
 
     @Override
     public int argNumber() {
-        return 1;
```
Comment (solonovamax) [inline common/src/main/java/com/dfsek/terra/api/structures/script/builders/ZeroArgFunctionBuilder.java:45]: lol

File: platforms/fabric/src/main/resources/terra.accesswidener [hunk 2]
```diff
@@ -1,3 +1,3 @@
 accessWidener v1 named
 
-extendable method net/minecraft/client/world/GeneratorType <init> (Ljava/lang/String;)V
```
Comment (solonovamax) [inline platforms/fabric/src/main/resources/terra.accesswidener:3]: what changed lol

**Review by solonovamax — CHANGES_REQUESTED** [review]
You may have forgotten some things

File: common/src/main/java/com/dfsek/terra/api/world/palette/holder/PaletteHolderBuilder.java [hunk 3]
```diff
@@ -17,8 +17,12 @@ public PaletteHolderBuilder add(int y, Palette<BlockData> palette) {
 
     @SuppressWarnings({"unchecked", "rawtypes", "RedundantSuppression"})
     public PaletteHolder build() {
-        Palette<BlockData>[] palettes = new Palette[paletteMap.lastKey() + 1];
-        for(int y = 0; y <= FastMath.max(paletteMap.lastKey(), 255); y++) {
+
+        int min = FastMath.min(paletteMap.keySet().stream().min(Integer::compareTo).orElse(0), 0);
```
[the remark this answers is not in the document]
Comment (dfsek) [reply common/src/main/java/com/dfsek/terra/api/world/palette/holder/PaletteHolderBuilder.java:21]: we dont have access to the world at this stage

File: platforms/fabric/src/main/resources/terra.accesswidener (hunk 2 above)
Comment (dfsek) [reply platforms/fabric/src/main/resources/terra.accesswidener:3]: the constructor was made `protected` so we can extend the class normally now

Comment (solonovamax) [reply platforms/fabric/src/main/resources/terra.accesswidener:3]: what I'm saying is, it removed the line and then added the exact same one.
image

**Review by solonovamax — APPROVED** [review]
All good.

**Pull request merged**